File: 가장 먼 노드.swift
import Foundation

var queue: [Int] = [1] // 몇 번 노드인지?
var connect: [[Int]] = []
var visited: [Bool] = []
var result: [Int] = []

func solution(_ n:Int, _ edge:[[Int]]) -> Int {
    
    connect = Array(repeating: [], count: n+1) // 0~n
    visited = Array(repeating: false, count: n+1)
    result = Array(repeating: 0, count: n+1)
    
    for ed in edge {
        let a = ed[0]
        let b = ed[1]
        connect[a].append(b)
        connect[b].append(a)
    }
    
    visited[1] = true
    
    while !queue.isEmpty {
        let q = queue.removeFirst() // 현재 노드
        bfs(q,result[q])
    }
    
    // print(result)
    let maxV = result.max()
    return result.filter { $0 == maxV}.count
    // result.filter { $0 == result.max()}.count로 했더니 틀림(할때마다 max를 계산하게 한다)
}

func bfs(_ f: Int, _ depth: Int) {
    let edges: [Int] = connect[f].filter { visited[$0] == false } // f에서 갈 수 있는 리스트
    
    for edge in edges {
        // if visited[edge] == false {
            queue.append(edge)
            visited[edge] = true
            result[edge] = depth+1
        // } 
    }
    
    
}

File: 가장 먼 노드.py
from collections import defaultdict


def solution(n, edges):
    answer = [0 for _ in range(n)]

    global maxDistance
    maxDistance = 0
    connected = defaultdict(list)  # i 입장에서 연결이 된 모음
    visited = [False for _ in range(n)]

    for edge in edges:
        a, b = edge[0] - 1, edge[1] - 1
        connected[a].append(b)
        connected[b].append(a)

    def bfs(k, visited):
        queue = []
        queue.append((0, 0))
        global maxDistance

        while queue:
            q, l = queue.pop(0)
            visited[q] = True

            for c in connected[q]:  # q와 연결된 애들
                if visited[c] == False:
                    visited[c] = True
                    queue.append((c, l+1))
                    maxDistance = max(maxDistance, l+1)
                    answer[c] = l+1

    bfs(0, visited)

    return answer.count(maxDistance)
